pkg/apiserver/server.go

```diff
@@ -183,42 +182,26 @@ func generateKotsadmID() error {
 		return nil
 	}
 	clusterID := clusters[0].ClusterID
-
-	// Write a Query to set/get an Event from the Store
-	isKotsadmIDGenerated, err := store.GetStore().IsKotsadmIDGenerated()
+	isKotsadmIDInDB, err := store.GetStore().IsKotsadmIDGenerated()
 	if err != nil {
-		return errors.Wrap(err, "Failed to check kotsadm id generation")
+		return err
 	}
-
-	// if the key exists, likely a fresh Install
-	if isKotsadmIDGenerated {
-		exists, err := s3pg.IsKotsadmIDConfigMapPresent()
-		if err != nil {
-			return errors.Wrap(err, "config map check error")
-		}
-		if exists {
-			// do nothing
-			return nil
-		}
-		//generate guid and use that as clusterId to identify that as a different install
-		clusterID = ksuid.New().String()
-		_, err = s3pg.CreateAdminIDConfigMap(clusterID)
-		if err != nil {
-			return errors.Wrap(err, "failed to to create config map")
-		}
+	cmpExists, err := s3pg.IsKotsadmIDConfigMapPresent()
+	if err != nil {
+		return err
 	}
-
-	// if the key does not exist, likely a Restore or pod restart
-	if !isKotsadmIDGenerated {
-		_, err := s3pg.CreateAdminIDConfigMap(clusterID)
-		if err != nil {
-			return errors.Wrap(err, "failed to create admin id")
-		}
-		// write to the db at the very if configmap creation succeeds and no other failures
+	if isKotsadmIDInDB && !cmpExists {
+		kotsadmID := ksuid.New().String()
+		err = s3pg.CreateAdminIDConfigMap(kotsadmID)
+	} else if !isKotsadmIDInDB && !cmpExists {
+		err = s3pg.CreateAdminIDConfigMap(clusterID)
+	} else if !isKotsadmIDInDB && cmpExists {
+		err = s3pg.UpdateAdminIDConfigMap(clusterID)
+	} else {
+		// id exists and so as configmap, noop
+	}
+	if err == nil {
 		err = store.GetStore().SetIsKotsadmIDGenerated()
-		if err != nil {
-			return errors.Wrap(err, "failed to set admin event status")
-		}
 	}
-	return nil
+	return err
```

Comment:
it's gonna be hard to know where the error came from, it would be better to wrap the error and return it after each function instead